hubspot-style/src/main/java/com/hubspot/immutables/utils/WireSafeEnum.java

```diff
@@ -0,0 +1,110 @@
+package com.hubspot.immutables.utils;
+
+import java.io.IOException;
+import java.util.Optional;
+
+import com.fasterxml.jackson.annotation.JsonValue;
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.core.JsonToken;
+import com.fasterxml.jackson.databind.BeanProperty;
+import com.fasterxml.jackson.databind.DeserializationContext;
+import com.fasterxml.jackson.databind.JavaType;
+import com.fasterxml.jackson.databind.JsonDeserializer;
+import com.fasterxml.jackson.databind.JsonMappingException;
+import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
+import com.fasterxml.jackson.databind.deser.ContextualDeserializer;
+import com.hubspot.immutables.utils.WireSafeEnum.Deserializer;
+
+@JsonDeserialize(using = Deserializer.class)
+public final class WireSafeEnum<T extends Enum<T>> {
+  private final Class<T> enumType;
+  private final String stringValue;
+  private final Optional<T> enumValue;
+
+  private WireSafeEnum(Class<T> enumType, String stringValue) {
+    this.enumType = enumType;
+    this.stringValue = stringValue;
+    this.enumValue = tryParse(enumType, stringValue);
+  }
+
+  @SuppressWarnings("unchecked")
+  private WireSafeEnum(T value) {
+    this.enumType = (Class<T>) value.getClass();
+    this.stringValue = value.name();
+    this.enumValue = Optional.of(value);
+  }
+
+  public static <T extends Enum<T>> WireSafeEnum<T> of(T value) {
+    return new WireSafeEnum<>(value);
+  }
+
+  public static <T extends Enum<T>> WireSafeEnum<T> of(String value, Class<T> enumType) {
```

Comment:
nit: I think `enumType` should come first in this method (it's first in the constructor)

Reply:
I thought it read better the other way around, but I don't feel strongly about it

Reply:
I guess most code usages will probably be of the variant that takes a literal enum constant anyway

Reply:
I think that class first is more consistent with the other ordering in this file and lines up with what's in [Guava's Enums](https://guava.dev/releases/19.0/api/docs/index.html?com/google/common/base/Enums.html) (though the semantics are different). I agree that it doesn't really matter.